In [18]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder

In [19]:
dataset=pd.read_csv('titanic_data (1).csv')
dataset.head(5)

PassengerId  Survived  Pclass  \
0            1       0.0       3   
1            2       1.0       1   
2            3       1.0       3   
3            4       1.0       1   
4            5       0.0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [20]:
dataset.isnull().sum()

PassengerId       0
Survived        418
Pclass            0
Name              0
Sex               0
Age             263
SibSp             0
Parch             0
Ticket            0
Fare              1
Cabin          1014
Embarked          2
dtype: int64

In [21]:
dataset[dataset.duplicated()]


Empty DataFrame
Columns: [PassengerId, Survived, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked]
Index: []

In [22]:
dataset=dataset.drop_duplicates()

In [23]:
dataset.duplicated().sum()

0

In [24]:
dataset = dataset.drop('Cabin' , axis=1)

In [25]:
print(dataset.isna().sum())


dataset=dataset.dropna(inplace=False)  

print(dataset.isna().sum())

PassengerId      0
Survived       418
Pclass           0
Name             0
Sex              0
Age            263
SibSp            0
Parch            0
Ticket           0
Fare             1
Embarked         2
dtype: int64
PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64


In [26]:
dataset= dataset.drop('Name' , axis=1)
dataset= dataset.drop('PassengerId' , axis=1)
dataset= dataset.drop('Ticket' , axis=1)
l = LabelEncoder()
dataset['Sex'] = l.fit_transform(dataset['Sex'])
dataset['Embarked'] = l.fit_transform(dataset['Embarked'])
dataset

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked
0         0.0       3    1  22.0      1      0   7.2500         2
1         1.0       1    0  38.0      1      0  71.2833         0
2         1.0       3    0  26.0      0      0   7.9250         2
3         1.0       1    0  35.0      1      0  53.1000         2
4         0.0       3    1  35.0      0      0   8.0500         2
..        ...     ...  ...   ...    ...    ...      ...       ...
885       0.0       3    0  39.0      0      5  29.1250         1
886       0.0       2    1  27.0      0      0  13.0000         2
887       1.0       1    0  19.0      0      0  30.0000         2
889       1.0       1    1  26.0      0      0  30.0000         0
890       0.0       3    1  32.0      0      0   7.7500         1

[712 rows x 8 columns]

In [27]:
from sklearn.preprocessing import MinMaxScaler

# Sample dataset (replace with your actual data)

dataset = pd.DataFrame(dataset)

# Create the MinMaxScaler
scaler = MinMaxScaler(feature_range=(0, 5))

# Fit and transform the dataset
dataset = scaler.fit_transform(dataset)



dataset = pd.DataFrame(dataset, columns=['Survived','Pclass','Sex','Age','SibSp','Parch','Fare','Embarked'])

# Now you can use .head() on X_train_scaled_df
dataset.head(3)


Survived  Pclass  Sex       Age  SibSp  Parch      Fare  Embarked
0       0.0     5.0  5.0  1.355868    1.0    0.0  0.070755       5.0
1       5.0     0.0  0.0  2.361146    1.0    0.0  0.695679       0.0
2       5.0     5.0  0.0  1.607188    0.0    0.0  0.077343       5.0

In [28]:
dataset1=dataset
dataset2=dataset

In [29]:
from sklearn.model_selection import train_test_split
y = dataset1['Survived']
X = dataset1.drop(columns=['Survived'])
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2)

In [30]:
from sklearn.linear_model import Lasso

linlasso = Lasso(alpha=0.001)
linlasso.fit(X_train, y_train)

print('lasso regression linear model intercept: {}'
     .format(linlasso.intercept_))
print('lasso regression linear model coeff:\n{}'
     .format(linlasso.coef_))
print('Non-zero features: {}'
     .format(np.sum(linlasso.coef_ != 0)))
print('R-squared score (training): {:.3f}'
     .format(linlasso.score(X_train, y_train)))
print('R-squared score (test): {:.3f}\n'
     .format(linlasso.score(X_test, y_test)))


lasso regression linear model intercept: 6.257022840817949
lasso regression linear model coeff:
[-0.41757202 -0.48690434 -0.51149513 -0.27106576 -0.07576542  0.04469344
 -0.09130544]
Non-zero features: 7
R-squared score (training): 0.431
R-squared score (test): 0.259



In [31]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score,classification_report, confusion_matrix

clf=DecisionTreeClassifier()
clf.fit(X_train, y_train)

# Predict on the test set
y_pred = clf.predict(X_test)

# Evaluate the model's performance
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)
print("Classification Report:\n", classification_report(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))

Accuracy: 0.8041958041958042
Classification Report:
               precision    recall  f1-score   support

         0.0       0.77      0.89      0.83        76
         5.0       0.85      0.70      0.77        67

    accuracy                           0.80       143
   macro avg       0.81      0.80      0.80       143
weighted avg       0.81      0.80      0.80       143

Confusion Matrix:
 [[68  8]
 [20 47]]


In [32]:
y1 = dataset2['Survived']
X1 = dataset2.drop(columns=['Survived'])
X_train1, X_test1, y_train1, y_test1 = train_test_split(X1, y1, test_size = 0.2)

In [33]:
from sklearn.linear_model import Ridge

linridge = Ridge(alpha=0.2)
linridge.fit(X_train1, y_train1)

print('ridge regression linear model intercept: {}'
     .format(linridge.intercept_))
print('ridge regression linear model coeff:\n{}'
     .format(linridge.coef_))
print('R-squared score (training): {:.3f}'
     .format(linridge.score(X_train, y_train)))
print('R-squared score (test): {:.3f}'
     .format(linridge.score(X_test, y_test)))
print('Number of non-zero features: {}'
     .format(np.sum(linridge.coef_ != 0)))


ridge regression linear model intercept: 6.123032699120971
ridge regression linear model coeff:
[-0.39086706 -0.49238542 -0.50508294 -0.28957695  0.01269577  0.03014715
 -0.08220666]
R-squared score (training): 0.430
R-squared score (test): 0.268
Number of non-zero features: 7


In [34]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score,classification_report, confusion_matrix

clf1=DecisionTreeClassifier()
clf1.fit(X_train1, y_train1)

# Predict on the test set
y_pred1 = clf1.predict(X_test1)

# Evaluate the model's performance
accuracy1 = accuracy_score(y_test1, y_pred1)
print("Accuracy:", accuracy1)
print("Classification Report:\n", classification_report(y_test1, y_pred1))
print("Confusion Matrix:\n", confusion_matrix(y_test1, y_pred1))

Accuracy: 0.7832167832167832
Classification Report:
               precision    recall  f1-score   support

         0.0       0.80      0.86      0.83        86
         5.0       0.76      0.67      0.71        57

    accuracy                           0.78       143
   macro avg       0.78      0.76      0.77       143
weighted avg       0.78      0.78      0.78       143

Confusion Matrix:
 [[74 12]
 [19 38]]


in my data set : the first model is better
the choice between these two models depends on the specific problem and the trade-offs you want to make. If you prioritize overall accuracy, the first model (with an accuracy of 80.42%) might be preferred. and if you want to focus on correctly identifying positive instances (class 5.0), the first model (with higher recall) could be better.